### PR TITLE
usm: Added protocol dispatcher

### PR DESCRIPTION
**Pull request by guyarb**
### What does this PR do?

Adds protocol classification to USM
Now when a packet will arrive USM's docket filter, we will classify the protocol and dispatch another socket filter program as a tail call to process the packet.

Currently, USM holds a dedicated map for the classification
In the future we should use a single map shared between USM and NPM

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by hmahmood — APPROVED** [review]

**Review by usamasaqib — APPROVED** [review]

File: pkg/network/ebpf/c/protocols/https.h [hunk 1]
```diff
@@ -28,6 +30,25 @@ static __always_inline void https_process(conn_tuple_t *t, void *buffer, size_t
...
+
+    protocol_t *cur_fragment_protocol_ptr = bpf_map_lookup_elem(&dispatcher_connection_protocol, &http.tup);
+    if (cur_fragment_protocol_ptr == NULL) {
+        protocol_t cur_fragment_protocol = PROTOCOL_UNKNOWN;
+        conn_tuple_t inverse_conn_tup = http.tup;
+        flip_tuple(&inverse_conn_tup);
+
+        cur_fragment_protocol_ptr = bpf_map_lookup_elem(&dispatcher_connection_protocol, &inverse_conn_tup);
+
+        // If we've already identified the protocol of the socket, no need to read the buffer and try to classify it.
```
Comment (val06) [inline pkg/network/ebpf/c/protocols/https.h:42]: rephrase the comment to match the code flow, now it "compliments" the code flow that is actually doesn't present in the code.

e.g: "try classifying the protocol if no prior identification exists"

File: pkg/network/ebpf/c/protocols/https.h [hunk 2]
```diff
@@ -28,6 +30,25 @@ static __always_inline void https_process(conn_tuple_t *t, void *buffer, size_t
     read_into_buffer(http.request_fragment, buffer, len);
     http.owned_by_src_port = http.tup.sport;
     log_debug("https_process: htx=%llx sport=%d\n", &http, http.owned_by_src_port);
+
+    protocol_t *cur_fragment_protocol_ptr = bpf_map_lookup_elem(&dispatcher_connection_protocol, &http.tup);
+    if (cur_fragment_protocol_ptr == NULL) {
+        protocol_t cur_fragment_protocol = PROTOCOL_UNKNOWN;
+        conn_tuple_t inverse_conn_tup = http.tup;
+        flip_tuple(&inverse_conn_tup);
```
Comment (val06) [inline pkg/network/ebpf/c/protocols/https.h:38]: can you explain why do you flip the conn tuple?

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 3]
```diff
@@ -84,15 +84,14 @@ static __always_inline void classify_protocol(protocol_t *protocol, const char *
...
 
-// Decides if the protocol_classifier should process the packet. We process not empty TCP packets.
-static __always_inline bool should_process_packet(struct __sk_buff *skb, skb_info_t *skb_info, conn_tuple_t *tup) {
-    // we're only interested in TCP traffic
-    if (!(tup->metadata & CONN_TYPE_TCP)) {
-        return false;
-    }
+// Returns true if the packet is TCP.
+static __always_inline bool is_tcp(conn_tuple_t *tup) {
+    return (tup->metadata&CONN_TYPE_TCP);
```
Comment (val06) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:89]: spaces

File: pkg/network/ebpf/c/protocols/protocol-dispatcher-helpers.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+#include <linux/types.h>
+
+#include "protocol-dispatcher-maps.h"
+#include "protocol-classification-helpers.h"
+#include "ip.h"
+
+// Returns true if the payload represents a TCP termination by checking if the tcp flags contains TCPHDR_FIN or TCPHDR_RST.
+static __always_inline bool is_tcp_termination(skb_info_t *skb_info) {
+    return skb_info->tcp_flags&(TCPHDR_FIN|TCPHDR_RST);
```
Comment (val06) [inline pkg/network/ebpf/c/protocols/protocol-dispatcher-helpers.h:12]: spaces

File: pkg/network/ebpf/c/protocols/protocol-dispatcher-helpers.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+static __always_inline void protocol_dispatcher_entrypoint(struct __sk_buff *skb) {
+    skb_info_t skb_info = {0};
+    conn_tuple_t skb_tup = {0};
+
+    // Exporting the conn tuple from the skb, alongside couple of relevant fields from the skb.
+    if (!read_conn_tuple_skb(skb, &skb_info, &skb_tup)) {
+        return;
+    }
+
+    // We handle the payload if it is an non empty TCP packet, or termination TCP packet.
```
Comment (val06) [inline pkg/network/ebpf/c/protocols/protocol-dispatcher-helpers.h:25]: same here, rephrase the comment to support the code.. (btw currently the comment doesn't match the condition)

e.g: "we don't process non tcp packets, nor empty tcp packets which are not tcp termination packets"

**Review by val06 — CHANGES_REQUESTED** [review]
added minor notes about comments and format

File: pkg/network/ebpf/c/protocols/https.h (hunk 2 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/https.h:38]: If we haven't identified the protocol on our direction (haven't done that yet, or could not do at all), then we should try and see if we were able to classify the opposite direction

for example, "binary" protocols (gRPC, kafka, etc.) that can be identified in only one direction

**Review by val06 — APPROVED** [review]

**Pull request merged**